src/main.rs

```diff
@@ -395,19 +419,36 @@ fn build_graph<'a>(
             };
             for dep in it {
                 let dep_idx = match graph.nodes.entry(dep_id) {
-                    Entry::Occupied(e) => *e.get(),
+                    Entry::Occupied(e) => {
+                        let key = *e.get();
+
+                        let mut node = &mut graph.graph[key];
+                        node.is_duplicate = true;
+
+                        key
+                    }
                     Entry::Vacant(e) => {
-                        pending.push(dep_id);
+                        if let Some(depth) = max_depth {
```

Comment:
Technically this check doesn't need to happen because it is taken care of in `print_tree()`, but it speeds up the common case of just printing the tree (not checking duplicates or starting from an arbitrary package) when a depth limit is specified.